token/program-2022/src/processor.rs

```diff
@@ -832,13 +871,27 @@ impl Processor {
                 return Err(TokenError::MintDecimalsMismatch.into());
             }
         }
+        let maybe_permanent_delegate = mint
+            .get_extension::<PermanentDelegate>()
+            .map(|e| Option::<Pubkey>::from(e.delegate))
+            .ok()
+            .flatten();
```

Comment:
oops, this is because i suggested `and_then` in the first instance in an offhand "its cool these are logically equivalent" way rather than a "you should make this change" way and so didnt flag the second instance

Reply:
So, I did the DRY bit, which ended up being pretty gnarly, since one call is on `StateWithExtensions` and the other is on `StateWithExtensionsMut`, which required a trait, which I then needed to use everywhere.  Let me know what you think!

Reply:
Sorry that was so much plumbing, but I actually like it quite a lot! I feel like I should have thought of those State implementation changes a while ago...

Reply:
Yeah, it ended up being really neat at the base level :smile: